haskell/test/Test/Ligo/BaseDAO/Proposal/Quorum.hs

```diff
@@ -20,17 +25,216 @@ import Test.Ligo.BaseDAO.Proposal.Config
 
 {-# ANN module ("HLint: ignore Reduce duplication" :: Text) #-}
 
+calculateThreshold :: Natural -> Natural -> GovernanceTotalSupply -> Integer -> QuorumThreshold -> QuorumThreshold
+calculateThreshold mcp cp (GovernanceTotalSupply gts) staked oldQt =
+  let
+    changePercent = QuorumThreshold $ percentageToFractionNumerator cp
+    participation = fractionToNumerator staked (fromIntegral gts)
+    maxChangePercent = QuorumThreshold $ percentageToFractionNumerator mcp
+    possibeNewQuorum = addF (subF oldQt (mulF oldQt changePercent)) (mulF participation changePercent)
+    one' = mkQuorumThreshold 1 1
+    maxNewQuorum = mulF oldQt (addF one' maxChangePercent)
+    minNewQuorum = divF oldQt (addF one' maxChangePercent)
+    -- old_quorum - (old_quorum * changePercent) + participation * changePercent
+    configMinQuorum = QuorumThreshold $ percentageToFractionNumerator 1
+    configMaxQuorum = QuorumThreshold $ percentageToFractionNumerator 99
+  in boundQt configMinQuorum configMaxQuorum (boundQt minNewQuorum maxNewQuorum possibeNewQuorum)
+  where
+    boundQt (QuorumThreshold qmin) (QuorumThreshold qmax) (QuorumThreshold qt) = QuorumThreshold $ min qmax (max qmin qt)
+    mulF (QuorumThreshold n1) (QuorumThreshold n2) = QuorumThreshold (div (n1 * n2) (fromIntegral fractionDenominator))
+    divF (QuorumThreshold n1) (QuorumThreshold n2) = QuorumThreshold (div (n1 * (fromIntegral fractionDenominator)) n2)
+    addF (QuorumThreshold n1) (QuorumThreshold n2) = QuorumThreshold (n1 + n2)
+    subF (QuorumThreshold n1) (QuorumThreshold n2) = QuorumThreshold (n1 - n2)
+    fractionToNumerator = mkQuorumThreshold
+
+checkQuorumThresholdDynamicUpdate
+  :: forall caps base m. (MonadEmulated caps base m, HasCallStack)
+  => (ConfigDesc Config -> OriginateFn m) -> m ()
+checkQuorumThresholdDynamicUpdate originateFn = do
```

Comment:
I'm taking this as an example, but it applies to all of the scenarios you converted.

So here you have restricted (by the `` constraint) the type of the scenario to only work on the emulated env and to fetch the value from the storage you do:
```haskell
  storage <- fromVal @FullStorage <$> getStorage' @(ToT FullStorage) dodDao
  let quorumThreshold= sQuorumThresholdAtCycle $ fsStorage storage
```

This works okay, but we have had to deal with this before and used a different solution, which is to define scenarios that are **not** restricted to be emulated by type, but instead accept a function to get the part of the storage that we need (depending if on emulated or network).

Take as example the `flushAcceptedProposals` scenario:
- it uses the more generic `MonadNettest` constraint
- it takes a `GetTotalSupplyFn` because it needs the total supply
- it's used with:
  ```haskell
  nettestScenarioOnEmulatorCaps <...> $ flushAcceptedProposals <...> getTotalSupplyEmulator
  ```

What are the advantages of this approach:
1. We can easily run the same scenario on the network as well (we don't because it takes too long)
   by providing a `GetTotalSupplyFn` that works on the network:
   ```haskell
   nettestScenarioOnNetworkCaps <...> $ flushAcceptedProposals <...> getTotalSupplyNetwork
   ```
2. we don't need to repeat the fetching code, the scenario just uses the `getTotalSupplyFn` it received and the actual implementation for the "get" is contained in the `Test.Ligo.BaseDAO.Common.StorageHelper` module.
   In fact, there you can see that [getFullStorage](https://github.com/tqtezos/baseDAO/blob/master/haskell/test/Test/Ligo/BaseDAO/Common/StorageHelper.hs#L26) has the same definition based on `getStorage'` that you use.

So I propose we do the same for the tests that you added here.

For example, with this scenario we'll need to:
1. add more helpers to `Test.Ligo.BaseDAO.Common.StorageHelper`:
   ```haskell
   ------------------------------------------------------------------------
   -- GetQuorumThresholdAtCycle
   ------------------------------------------------------------------------
   
   type GetQuorumThresholdAtCycleFn m = Address -> m QuorumThresholdAtCycle
   
   getQuorumThresholdAtCycleFnEmulator :: Address -> EmulatedT PureM QuorumThresholdAtCycle
   getQuorumThresholdAtCycleFnEmulator addr =
     sQuorumThresholdAtCycle . fsStorage <$> getFullStorage addr
   ```
2. modify the scenario type to take the helper as an argument:
   ```suggestion
   checkQuorumThresholdDynamicUpdate
     :: (MonadNettest caps base m, HasCallStack)
     => (ConfigDesc Config -> OriginateFn m)
     -> GetQuorumThresholdAtCycleFn m
     -> m ()
   checkQuorumThresholdDynamicUpdate originateFn getQuorumFn = do
   ```
3. use the given function to fetch the storage fieds, so replace:
   ```haskell
     storage <- fromVal @FullStorage <$> getStorage' @(ToT FullStorage) dodDao
     let quorumThresholdActual_ = sQuorumThresholdAtCycle $ fsStorage storage
   ```
   with
   ```haskell
     quorumThresholdActual_ <- getQuorumFn $ unTAddress dodDao
   ```
4. call the scenario with:
   ```haskell
         [ nettestScenarioOnEmulatorCaps "updates quorum-threshold correctly" $
             checkQuorumThresholdDynamicUpdate (originateLigoDaoWithConfigDesc dynRecUnsafe) 
               getQuorumThresholdAtCycleFnEmulator
   ```
What do you say? Does that sound reasonable to you?

Reply:
Have included this.